### PR TITLE
Release 1.1.23

### PR DESCRIPTION
**Pull request by anonymoussprocket**
- [x] support for dexter & quipu swap token purchases & sales
- [x] support for sending HEN NFTs
- [x] initial support for PLENTY, KALAM, hDAO, WRAP, BTCtz, Flame, Hedgehodge, USDS, SMAK tokens
- [x] improved tzBTC token rendering
- [x] various updates to Beacon operations processing

### CODE REVIEW

File: src/contracts/HicNFT/components/Collection/gallery.tsx [hunk 1]
```diff
@@ -17,7 +17,17 @@ const ArtGallery = (props: ArtGalleryProps) => {
...
-        return <ArtPiece key={p.piece} objectId={p.piece} amount={p.amount} price={p.price} receivedOn={p.receivedOn} action={p.action} />;
+        return (
+            <ArtPiece
+                key={p.piece}
+                objectId={p.piece}
+                amount={p.amount}
+                price={p.price}
+                receivedOn={p.receivedOn}
+                action={p.action}
+                transferTrigger={undefined}
```
Comment (dariuszsobkowicz) [inline src/contracts/HicNFT/components/Collection/gallery.tsx]: `transferTrigger` unused

File: src/contracts/HicNFT/components/TransferModal/TransferModal.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                                    onChange={(val) => setPassPhrase(val)}
+                                    containerStyle={{ width: '60%', marginTop: '10px' }}
+                                />
+                            )}
+                            <InvokeButton buttonTheme="primary" disabled={isDisabled} onClick={onTransfer}>
+                                {t('components.hicNFT.modal.send')}
+                            </InvokeButton>
+                        </PasswordContainer>
+                    </Container>
+                    {/* <Loader /> */}
```
Comment (dariuszsobkowicz) [inline src/contracts/HicNFT/components/TransferModal/TransferModal.tsx:107]: add todo to clean up unused or improve a component

File: src/contracts/Plenty/components/BalanceBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        <Container>
+            <TopRow isReady={isReady}>
+                <Breadcrumbs>{breadcrumbs}</Breadcrumbs>
+                <Update onClick={onSyncWallet} time={time} isReady={isReady} isWalletSyncing={isWalletSyncing} />
+            </TopRow>
+            <BottomRow isReady={isReady}>
+                <AddressInfo>
+                    <AddressTitle>{displayName}</AddressTitle>{' '}
+                    {token.helpLink && (
+                        <>
```
Comment (dariuszsobkowicz) [inline src/contracts/Plenty/components/BalanceBanner.tsx]: no need `<>` for a single component

File: src/contracts/Plenty/components/Swap.tsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                quipuCost = getXTZBuyExchangeRate(
+                    new BigNumber(tokenValue).multipliedBy(10 ** (token.scale || 0)).toString(),
+                    quipuState.tokenBalance,
+                    quipuState.coinBalance
+                );
+            }
+
+            setQuipuTokenCost(applyFees(quipuCost.xtzAmount, 'buy'));
+            setBestMarket('QuipuSwap');
+        } else if (side === 'sell') {
```
Comment (dariuszsobkowicz) [inline src/contracts/Plenty/components/Swap.tsx:127]: buy and sell if looks almost the same so code can be reduced

File: src/contracts/Plenty/components/Swap.tsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    return (
+        <Container>
+            <RowContainer>
+                <div style={{ width: '100%', paddingRight: '10px' }}>
+                    <SegmentedControlContainer>
+                        <RestoreTabs type={tradeSide} changeFunc={(val) => onTypeChange(val)} />
+                    </SegmentedControlContainer>
+                </div>
+                <ColumnContainer>
+                    <div style={{ width: '100%', paddingLeft: '10px' }}>
```
Comment (dariuszsobkowicz) [inline src/contracts/Plenty/components/Swap.tsx:193]: used twice so can be a single component

File: src/contracts/Plenty/container.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                const itemsCount = 5;
+                const pageCount = Math.ceil(processedTransactions.length / itemsCount);
+
+                const firstNumber = (currentPage - 1) * itemsCount;
+                const lastNumber = Math.min(currentPage * itemsCount, processedTransactions.length);
+
+                const transactionSlice = processedTransactions.slice(firstNumber, lastNumber);
+
+                return (
+                    <Fragment>
```
Comment (dariuszsobkowicz) [inline src/contracts/Plenty/container.tsx]: <> will work

File: src/contracts/Token2Contract/container.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                const itemsCount = 5;
+                const pageCount = Math.ceil(processedTransactions.length / itemsCount);
+
+                const firstNumber = (currentPage - 1) * itemsCount;
+                const lastNumber = Math.min(currentPage * itemsCount, processedTransactions.length);
+
+                const transactionSlice = processedTransactions.slice(firstNumber, lastNumber);
+
+                return (
+                    <Fragment>
```
Comment (dariuszsobkowicz) [inline src/contracts/Token2Contract/container.tsx]: <>

File: src/contracts/components/Swap/index.tsx [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    return (
+        <Container>
+            <RowContainer>
+                <div style={{ width: '100%', paddingRight: '10px' }}>
+                    <SegmentedControlContainer>
+                        <RestoreTabs type={tradeSide} changeFunc={(val) => onTypeChange(val)} />
+                    </SegmentedControlContainer>
+                </div>
+                <ColumnContainer>
+                    <div style={{ width: '100%', paddingLeft: '10px' }}>
```
Comment (dariuszsobkowicz) [inline src/contracts/components/Swap/index.tsx:276]: can be added and imported from styles

File: src/contracts/components/Swap/util.ts [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+            params,
+            TezosParameterFormat.Micheline,
+            TezosConstants.HeadBranchOffset,
+            true
+        );
+
+        return r.operationGroupID.replace(/\\|"|\n|\r/g, '');
+    } catch (err) {
+        console.log(`failed in sendDexterBuy ${JSON.stringify(err)}}`);
+        return undefined;
```
Comment (dariuszsobkowicz) [inline src/contracts/components/Swap/util.ts]: unnecessary

File: src/contracts/components/Swap/util.ts [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    );
+
+    try {
+        const opGroup = await TezosNodeWriter.prepareOperationGroup(tezosNode, keyStore, nextCounter - 1, [approveOp, sellOp], true);
+        const r = await TezosNodeWriter.sendOperation(tezosNode, opGroup, signer);
+
+        return r.operationGroupID.replace(/\\|"|\n|\r/g, '');
+    } catch (err) {
+        console.log(`failed in sendDexterSell ${JSON.stringify(err)}}`);
+        return undefined;
```
Comment (dariuszsobkowicz) [inline src/contracts/components/Swap/util.ts]: unnecessary

File: src/contracts/components/Swap/util.ts [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+            params,
+            TezosParameterFormat.Micheline,
+            TezosConstants.HeadBranchOffset,
+            true
+        );
+
+        return r.operationGroupID.replace(/\\|"|\n|\r/g, '');
+    } catch (err) {
+        console.log(`failed in sendQuipuBuy ${JSON.stringify(err)}}`);
+        return undefined;
```
Comment (dariuszsobkowicz) [inline src/contracts/components/Swap/util.ts]: unnecessary

File: src/contracts/components/Swap/util.ts [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+    );
+
+    try {
+        const opGroup = await TezosNodeWriter.prepareOperationGroup(tezosNode, keyStore, nextCounter - 1, [approveOp, sellOp], true);
+        const r = await TezosNodeWriter.sendOperation(tezosNode, opGroup, signer);
+
+        return r.operationGroupID.replace(/\\|"|\n|\r/g, '');
+    } catch (err) {
+        console.log(`failed in sendQuipuSell ${JSON.stringify(err)}}`);
+        return undefined;
```
Comment (dariuszsobkowicz) [inline src/contracts/components/Swap/util.ts]: unnecessary

**Review by dariuszsobkowicz — APPROVED** [review]
lgtm

**Review by vishakh — APPROVED** [review]

**Pull request merged**